File: Dart/quicksort.dart

class QuickSort {
  List<int> quick(List<int> array) {
    sort(array, 0, array.length - 1);
    return array;
  }

  void sort(List<int> array, int start, int end) {
    if (start >= end) {
      return;
    }
    int pivot = start;
    int left = start + 1;
    int right = end;
    while (left <= right) {
      if (array[left] > array[pivot] && array[right] < array[pivot]) {
        swap(array, left, right);
        left++;
        right--;
      }
      if (array[left] <= array[pivot]) {
        left++;
      }
      if (array[right] >= array[pivot]) {
        right--;
      }
    }
    swap(array, pivot, right);
    sort(array, start, right - 1);
    sort(array, right + 1, end);
  }

  void swap(List<int> array, int i, int j) {
    int temp = array[i];
    array[i] = array[j];
    array[j] = temp;
  }
}

File: JS/quicksort.js
const array=[0,-90,0,88,2]

const quick=(arr)=>{
    if(arr.length<2){
        return arr
    }
    let pivot=arr[arr.length-1]
    let leftArray=[]
    let rightArray=[]
    for(let i=0;i<arr.length-1;i++){
        if(arr[i]<pivot){
            leftArray.push(arr[i])
        }else{
            rightArray.push(arr[i])
        }
    }
    return [...quick(leftArray),pivot,...quick(rightArray)]
}

console.log(quick(array))

// output:[ -90, 0, 0, 2, 88 ]